### PR TITLE
Backport HMACContext to 3.x

### PR DESCRIPTION
**Pull request by tavurth**
I backported this PR https://github.com/godotengine/godot/pull/43536

Thank you @jonbonazza!

I also tested with this simple script from the docs:

https://github.com/godotengine/godot-proposals/issues/1098

```gdscript
extends Control

var ctx = HMACContext.new()

func _ready():
	var key = "supersecret".to_utf8()
	var err = ctx.start(HashingContext.HASH_SHA256, key)
	assert(err == OK)
	var msg1 = "this is ".to_utf8()
	var msg2 = "vewy vewy secret".to_utf8()
	err = ctx.update(msg1)
	assert(err == OK)
	err = ctx.update(msg2)
	assert(err == OK)
	var hmac = ctx.finish()
	print(hmac.hex_encode())
```

https://github.com/godotengine/godot-proposals/issues/1127

### CODE REVIEW

**Review by Faless — CHANGES_REQUESTED** [review]
The script you mention in the PR is not the one for HMAContext, but for AESContext (which was already backported). 
The tests in `main/tests/*` cannot be run as they are not called in `main_test.cpp` and would not compile anyway due to missing references.
We can either skip those, or add the missing functions.

File: modules/mbedtls/crypto_mbedtls.h [hunk 1]
```diff
@@ -34,9 +34,9 @@
 #include "core/crypto/crypto.h"
 #include "core/resource.h"
 
-#include <mbedtls/ctr_drbg.h>
-#include <mbedtls/entropy.h>
-#include <mbedtls/ssl.h>
+#include "thirdparty/mbedtls/include/mbedtls/ctr_drbg.h"
+#include "thirdparty/mbedtls/include/mbedtls/entropy.h"
+#include "thirdparty/mbedtls/include/mbedtls/ssl.h"
```
Comment (Faless) [inline modules/mbedtls/crypto_mbedtls.h:39]: Why was this changed? This would break `buildin_mbedtls=no`

File: modules/mbedtls/register_types.cpp [hunk 2]
```diff
@@ -35,6 +35,10 @@
 #include "packet_peer_mbed_dtls.h"
 #include "stream_peer_mbedtls.h"
 
+#ifdef TESTS_ENABLED
```
Comment (Faless) [inline modules/mbedtls/register_types.cpp:38]: This does nothing in `3.x`, and the referenced header file is not present anyway. should be removed.

File: modules/mbedtls/crypto_mbedtls.h (hunk 1 above)
Comment (tavurth) [reply modules/mbedtls/crypto_mbedtls.h:39]: Checking this thank you!

Comment (tavurth) [reply modules/mbedtls/crypto_mbedtls.h:39]: This issue seems to be that:

```

[ 13%] In file included from main/tests/test_crypto.cpp:34:
./modules/mbedtls/crypto_mbedtls.h:37:10: fatal error: 'mbedtls/ctr_drbg.h' file not found
#include <mbedtls/ctr_drbg.h>
         ^~~~~~~~~~~~~~~~~~~~
```

File: modules/mbedtls/register_types.cpp (hunk 2 above)
Comment (tavurth) [reply modules/mbedtls/register_types.cpp:38]: Thanks!

File: modules/mbedtls/crypto_mbedtls.h (hunk 1 above)
Comment (Faless) [reply modules/mbedtls/crypto_mbedtls.h:39]: If you check the `master` branch, the main test does not (and should not) include the module.

Comment (tavurth) [reply modules/mbedtls/crypto_mbedtls.h:39]: I would rather add the tests if possible or you think they're unnecessary?
I will check the main branch!

Thanks for the review :)

Comment (Faless) [reply modules/mbedtls/crypto_mbedtls.h:39]: Make sure they run though, the approach to testing has completely changed between `3.x` and `master`.

**Comment by tavurth** [comment]
Thank you for the review @Faless!

I've made the changes you requested, since those tests won't run anyway (`/mbedtls/tests` I couldn't figure out the dependency tree) I removed them for now.

I couldn't find how to run tests locally, is that documented anywhere?

**Comment by Faless** [comment]
> I couldn't find how to run tests locally

`bin/godot --test crypto`

**Comment by tavurth** [comment]
Thank you!

```c++
                                                                                                                                                     15m ⚑
▶ ./bin/godot.osx.tools.x86_64 --test crypto
arguments
0: ./bin/godot.osx.tools.x86_64
1: --test
2: crypto
Current path: /Users/will/src/godot
Godot Engine v3.4.beta.custom_build.0f88bdc07 - https://godotengine.org
OpenGL ES 3.0 Renderer: Intel(R) Iris(TM) Plus Graphics 655
OpenGL ES Batching: ON

Registered camera FaceTime HD Camera (Built-in) with id 1 position 0 at index 0
	PASS

Passed 1 of 1 tests
```

File: main/tests/test_crypto.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "core/crypto/crypto.h"
+#include "core/os/os.h"
+
+namespace TestCrypto {
```
Comment (akien-mga) [inline main/tests/test_crypto.cpp:34]: Does the test pass when compiling with `module_mbedtls_enabled=no`?

Comment (tavurth) [reply main/tests/test_crypto.cpp:34]: Good question, I just compiled using:

```
scons platform=osx arch=x86_64 module_mbedtls_enabled=no --jobs=$(sysctl -n hw.logicalcpu)
```

And tests are passing, is that correct?

**Review by Faless — APPROVED** [review]
Looks good, thanks! :+1:

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution 🎉

**Comment by tavurth** [comment]
Thank you @akien-mga @Faless!
